### DISCUSSION TITLE
Can Autocomplete take arbitrary values?

### DISCUSSION THREAD

**Discussion by sebastian-eggers-pcg** [formkit/formkit#1422] [General]
Currently, Autocomplete always resets its value when the user starts typing and leaves the input without choosing a value.
In our usecase, we want to use Autocomplete only for suggestions, but also want to accept arbitrary inputs.
Is there an option for this (we did not find any), or can this somehow be solved somehow via an event callback or hook?

**Comment by sashamilenkovic** [comment]
@sebastian-eggers-pcg The taglist has a prop `allow-new-values` that will allow for arbitrary values to be entered on press of the Enter key. There are also modifications you can make to allow for other key commands to input a value. Is that something you're looking for? Or rather, would you want the text itself to become a value without having to press a key?

**Reply by sebastian-eggers-pcg** [reply]
What we are looking for is basically a solution for the following use case:

We use autocomplete inside an address form. The user enters his street, and is offered an autocomplete for known streets in his city. But sometimes the information in the autocomplete may not be up-to-date, so we want to allow the user to enter a value that is not inside the autocomplete values.

The best workaround I've come up so far is to always add the current input as label/value pair to the list of options. But the user still needs to actively select it.

The best solution for us would be that if the user leaves the autocomplete input, then the current input remains as value. Right now the value is always reverted to the last valid value (if there was one before, and the field has not been cleared completely) or cleared (any other case).

**Reply by sashamilenkovic** [reply]
Gotcha, so maybe something like this? https://formkit.link/63bc625116fabc60369e480e6128eeec. So on blur, if there is input text, the value will be emitted.

**Reply by sebastian-eggers-pcg** [reply]
We will try this, thanks for your help :)

**Reply by giteronet** [reply]
We have a little problem here. On your initial selection of an option, the system will not register the option chosen but will instead process the current user input. This issue can be overcome by verifying the absence of e.relatedTarget in the 'if' condition.

Reproduction example:
Type Ba and click on Banana => result: Ba
Next add an 'n' to 'Ba' and click on Banana again => now it works.

Solution:
You have to detect if the user clicked an item. We can do this with relatedTarget. It returns you the clicked list item. When you click on no item, then it will be null and we want to set the current user input as a value only if he didn't click a list item. I guess checking for `if (node.props.inputText && !e.relatedTarget)` is enough. I think even only checking for `!e.relatedTarget` is okay.

**Reply by sashamilenkovic** [reply]
@giteronet Ah, you're right. Sorry about that. Hmmm, I feel like this is a worthwhile feature. I'd try to produce another quick workaround but it'd be better to just have this implemented as a feature (probably just for single-select autocomplete). I'll convert this to an issue and mark it as an enhancement. 

Follow up question, would you prefer if the input text was set as the value as soon as the user began typing? Or only when a blur occurs?

**Reply by sashamilenkovic** [reply]
Enhancement Issue filed here: https://github.com/formkit/formkit/issues/1423

**Reply by giteronet** [reply]
Yes, I think as a feature would be great. Thank you! :)
Probably best as soon as the user began typing. Otherwise the value will probably also disappear when the user presses tab. At least in the example it is like that and to fix this, I also needed to listen to a keydown event.

**Comment by nhedger** [comment]
Also interested in this feature, thanks for creating the discussion!
